Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp

```diff
@@ -289,6 +289,48 @@ void RemoteGraphicsContextGLProxy::simulateEventForTesting(SimulatedEventForTest
     }
 }
 
+void RemoteGraphicsContextGLProxy::getBufferSubData(GCGLenum target, GCGLintptr offset, std::span<uint8_t> data)
+{
+    if (isContextLost())
+        return;
+
+    if (data.empty())
+        return;
+
+    static constexpr size_t getBufferSubDataInlineSizeLimit = 64 * KB; // NOTE: when changing, change the value in RemoteGraphicsContextGL too.
+
+    auto copyToData = [&](std::span<const uint8_t> replyData) {
+        memcpy(data.data(), replyData.data(), replyData.size_bytes());
+    };
+
+    if (data.size_bytes() > getBufferSubDataInlineSizeLimit) {
+        RefPtr<SharedMemory> replyBuffer = SharedMemory::allocate(data.size_bytes());
+        if (!replyBuffer)
+            goto inlineCase;
+        auto handle = replyBuffer->createHandle(SharedMemory::Protection::ReadWrite);
+        if (!handle)
+            goto inlineCase;
+        auto sendResult = sendSync(Messages::RemoteGraphicsContextGL::GetBufferSubDataSharedMemory(target, offset, WTFMove(*handle)));
+        if (!sendResult.succeeded()) {
+            markContextLost();
+            return;
+        }
+        auto [readBytes] = sendResult.takeReply();
+        if (!readBytes)
+            return;
+        copyToData(replyBuffer->span().subspan(0U, readBytes));
+        return;
+    }
+inlineCase:
+    auto sendResult = sendSync(Messages::RemoteGraphicsContextGL::GetBufferSubDataInline(target, offset, data.size_bytes()));
```

Comment:
if the SharedMemory allocation fails, you end up here with size limit bigger than what the receiver will allow. I wonder if this is a problem with texImage2D, too?

It's anyway good idea to iterate and transfer in blocks.

Reply:
This code is taken from readPixels, so it has the same problem?